### PR TITLE
 Manually delete physical files for ENR entries when leaving scope.

### PR DESCRIPTION
**Pull request by Sairakan**
### Description

Second attempt at https://github.com/babelfish-for-postgresql/babelfish_extensions/pull/2930 (which was reverted). This time, we completely ignore the pending deletes list, and instead manually trigger file deletion for each of the ENR entries we are dropping during ENRDropTempTables(). This ensures that we don't need to mess with the pendingDeletes list at all, while still properly deleting the physical storage for ENR entries when they are no longer in scope (such as at the end of a stored procedure call, explicit transaction, or batch).
 
### Issues Resolved

BABEL-4751
 
### Check List

- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the PostgreSQL license, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/postgresql_modified_for_babelfish/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/backend/utils/misc/queryenvironment.c [hunk 1]
```diff
@@ -1572,6 +1584,10 @@ ENRDropTempTables(QueryEnvironment *queryEnv)
 		object.objectSubId = 0;
 		object.objectId = enr->md.reliddesc;
 		add_exact_object_address(&object, objects);
+		/*
+		 * Also register the physical file to be deleted at the end of scope.
+		 */
+		ENRDelete(enr->md.reliddesc);
```
Comment (surendravishnoi) [inline src/backend/utils/misc/queryenvironment.c]: Why not delete the file right here rather than adding it to delete list and later calling smgrDoPendingDeletes. We can write a new function to delete a file in SMGR

Comment (Sairakan) [reply src/backend/utils/misc/queryenvironment.c]: Let me look into this. The original reasoning behind doing it this way is that I am not sure how much of the storage deletion needs to be copied, vs just immediately calling smgr_close() on the relation

Comment (Sairakan) [reply src/backend/utils/misc/queryenvironment.c]: I've implemented this suggestion - now the files are immediately cleaned up during this function, which has the added benefit of not needing to modify any other function signatures or anything else in the code.

File: src/backend/utils/misc/queryenvironment.c [hunk 2]
```diff
@@ -1572,6 +1578,45 @@ ENRDropTempTables(QueryEnvironment *queryEnv)
...
 		add_exact_object_address(&object, objects);
+		/*
+		 * Delete the physical storage for the relation.
+		 * See: smgrDoPendingDeletes()
+		 */
+		rel = relation_open(enr->md.reliddesc, AccessExclusiveLock);
+		srel = smgropen(rel->rd_locator, rel->rd_backend);
+
+		/* allocate the initial array, or extend it, if needed */
+		if (maxrels == 0)
```
Comment (surendravishnoi) [inline src/backend/utils/misc/queryenvironment.c:1589]: why are we doing this? we have single relationship, we should just pass pointer to that relation with count as 1 to smgrdounlinkall

Comment (Sairakan) [reply src/backend/utils/misc/queryenvironment.c:1589]: We do not necessarily only have one relation - there can be multiple table variables and temp tables defined in the same scope, which would all be cleaned up at once by this function.

**Review by surendravishnoi — APPROVED** [review]

**Review by lejaokri — APPROVED** [review]

File: src/backend/utils/misc/queryenvironment.c (hunk 2 above)
Comment (Sairakan) [reply src/backend/utils/misc/queryenvironment.c:1589]: Correction: currently table variables are being cleaned up prior to this function during pltsql_clean_table_variables(), which manually executes a DROP TABLE <tblname> statement to drop the table variable. Since that function only registers the physical file for deletion at transaction rather than immediately, we probably should remove that function and let this function do both table variable and temp table cleanup, but that can be done in a later commit.

TLDR: currently only temp tables in scope are being cleaned up here, and not table variables, which are handled separately.
